src/nccl_ofi_rdma.c

```diff
@@ -306,65 +263,77 @@ static void unlink_topo_file()
 static int write_topo_file(nccl_ofi_topo_t *topo)
 {
 	int ret = 0;
-	int rc = 0;
-	FILE *file;
-	char *filename;
-	int fd;
+	int topo_fd = -1;
+	int write_fd = -1;
+	FILE *file = NULL;
 
 	/* This function is a no-op in case writing topology file is not enabled explicitly */
 	if (!ofi_nccl_topo_file_write_enable()) {
+		NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+			       "Topology write not enabled; skipping");
 		goto exit;
 	}
 
-	nccl_net_ofi_mutex_lock(&topo_file_lock);
-
-	if (topo_file_unlink) {
-		/* A topology file has already been written and stored
-		 * such that it can be unlinked later. Do not write
-		 * another topology file since it would end up
-		 * overriding the stored filename. */
-		goto unlock;
+	/* If `NCCL_TOPO_FILE` is already set, don't set it again. */
+	if (getenv("NCCL_TOPO_FILE")) {
```

Comment:
This isn't what we talked about in slack.  You need to make sure this doesn't point to some parent process's memory, since the parent may exit.  I thought what we talked about was that if NCCL_TOPO_FILE is set, you would call realpath on it and compare for the expected `memfd:nccl-ofi-topo.xml` or whatever.

Reply:
After thinking about it, I don't see why this is necessary. File descriptors are copied to the child. So, if the parent process sets `NCCL_TOPO_FILE == /dev/fd/X`, then `fork()`s, the child will also have a topology file at `/dev/fd/X` that will be valid until the child exits. So we shouldn't need to modify `NCCL_TOPO_FILE` in this case.

For what it's worth, I tested this with a small example and confirmed that the child's descriptor is still valid after the parent exits. I also tested this PR with an ML app (Fairseq FSDP) which I believe has this forking behavior.

Reply:
Comment added.

Reply:
We talked about this in slack.

parent:
 - plugin_init()
 - fork()
 - exit()

child:
  - plugin_init()
  - compute

There's now a race -- if parent exists before child's nccl_init() is completed, then the file isn't there.

Reply:
Yes, I know we talked about this in Slack, but I'm not seeing the race here.

After parent calls `plugin_init()`, `NCCL_TOPO_FILE=/dev/fd/X` is a valid memfd topology file. At `fork()`, the file descriptor is copied, so that in the child, `NCCL_TOPO_FILE=/dev/fd/X` points to this same memfd topology file. The file stays valid, even after the parent exits, until all file descriptors referring to it are closed (I tested this to confirm), so it should be OK to use the same topology file path in the child.

I would prefer to avoid the complication around testing whether the topology file comes from the parent or the user, if it isn't necessary.

Reply:
Cleaner way to solve this? 

```
int get_topo_fd(void) {
  static int fdno = memfd_create("ofi_nccl_topo", 0);
  return fdno; 
}

[[gnu::constructor]]
void ensure_topo_fd(void) {
  (void)get_topo_fd();
}
```

Now there is no race on `plugin_init()`, it's initialized consistently by dlopen and inherited by all children.

Reply:
Discussed on Slack, decided to leave as-is for now.